### PR TITLE
🔒  7 Ego Login w/ Cookies

### PR DESCRIPTION
**Pull request by demariadaniel**
Adds additional Login functionality including token management and middleware handling
Experimental, more functionality than [#32](https://github.com/icgc-argo/rdpc-ui/pull/32) 

Covers both:
https://app.zenhub.com/workspaces/icgc-argo-platform-dk-production-board-5e542d38415f5034e9fed89d/issues/gh/icgc-argo/rdpc-ui/4

https://app.zenhub.com/workspaces/icgc-argo-platform-dk-production-board-5e542d38415f5034e9fed89d/issues/gh/icgc-argo/rdpc-ui/7

### CODE REVIEW

**Comment by ciaranschutte** [comment]
Looks great! A couple of suggestions and a couple of general points.

1) I think starting this project off fresh is a really good opportunity to fix anything we wanted to from `platform-ui` - take advantage of new apis, clean up code etc

2) structure of project. Thoughts on putting `hooks` into it's own folder? They're not really `global` in my mind. `egoJwt` looking great - thoughts on putting this under `lib`? As it's not really a `util` it's actually doing all the `Ego` work so it's closer to a `@/lib/ego`. 

3) I think it's in an upcoming ticket so I won't go into too much depth. I think we'll have to rejig the page layout a little on where we put <AuthProvider> as it may be difficult to update the context where it is now. No need to change anything yet.

File: src/global/hooks/auth.tsx [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+export default function AuthProvider({
+	authData,
+	children,
+}: {
+	authData: T_AuthContext;
+	children: ReactNode;
+}) {
+	return <AuthContext.Provider value={authData}>{children}</AuthContext.Provider>;
+}
```
Comment (ciaranschutte) [inline src/global/hooks/auth.tsx:29]: please add copyright

File: src/global/utils/egoJwt/index.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+const TokenUtils = createEgoUtils(getAppConfig().EGO_PUBLIC_KEY);
+
+export const decodeToken = memoize((egoJwt?: string) =>
+	egoJwt ? TokenUtils.decodeToken(egoJwt) : null,
+);
+
+export const isValidJwt = (egoJwt: string) => !!egoJwt && TokenUtils.isValidJwt(egoJwt);
+
+export const getPermissionsFromToken: (egoJwt: string) => string[] = (egoJwt) =>
+	isValidJwt(egoJwt) ? TokenUtils.getPermissionsFromToken(egoJwt) : [];
```
Comment (ciaranschutte) [inline src/global/utils/egoJwt/index.ts:15]: please add copyright

**Review by ciaranschutte — CHANGES_REQUESTED** [review]

**Comment by demariadaniel** [comment]
> Looks great! A couple of suggestions and a couple of general points.
> 
> 1. I think starting this project off fresh is a really good opportunity to fix anything we wanted to from `platform-ui` - take advantage of new apis, clean up code etc
> 2. structure of project. Thoughts on putting `hooks` into it's own folder? They're not really `global` in my mind. `egoJwt` looking great - thoughts on putting this under `lib`? As it's not really a `util` it's actually doing all the `Ego` work so it's closer to a `@/lib/ego`.
> 3. I think it's in an upcoming ticket so I won't go into too much depth. I think we'll have to rejig the page layout a little on where we put  as it may be difficult to update the context where it is now. No need to change anything yet.

Agreed, we should sit down and work out the structure for the app at this point!

Hooks -- these are for Auth, and generally everything uses Auth, so that does seem Global to me, but there's not exactly a lot of `hooks` there, so maybe use a different folder? Thoughts?
/lib /ego makes sense, will update

And agreed to Layout updates, definitely want our root App setup vs our UI Layout files. This branch is just adding stubs for those things so we have a POC Login/Logout.

File: src/global/utils/egoJwt/index.ts (hunk 2 above)
Comment (demariadaniel) [reply src/global/utils/egoJwt/index.ts:34]: Updated

File: src/global/hooks/auth.tsx (hunk 1 above)
Comment (demariadaniel) [reply src/global/hooks/auth.tsx:48]: Updated

File: src/global/hooks/auth.tsx [hunk 3]
```diff
@@ -0,0 +1,29 @@
+import { createContext, ReactNode } from 'react';
+import Cookies from 'js-cookie';
+import { EGO_JWT_KEY } from '../constants';
+
+type T_AuthContext = {
```
[the remark this answers is not in the document]
Comment (demariadaniel) [reply src/global/hooks/auth.tsx:24]: This is straight out of Platform UI, I could update to `AuthContextValue` so we have distinction btwn `AuthContext.Provider`, `AuthContext.Consumer` + `AuthContextValue`

Any style preferences here?

File: .env.schema [hunk 4]
```diff
@@ -0,0 +1,10 @@
+#
+# Defaults for all environments
+#
+
+NEXT_PUBLIC_TEST=RDPC
+
+######### Ego
+EGO_API_ROOT=http://localhost:8081
+EGO_CLIENT_ID=rdpc-ui-local
+EGO_PUBLIC_KEY=
```
[the remark this answers is not in the document]
Comment (demariadaniel) [reply .env.schema:10]: Aha I missed this, let me give a read
I was concerned about committing an actual env, included env.schema as a skeleton

File: src/global/utils/egoJwt/index.ts [hunk 5]
```diff
@@ -0,0 +1,15 @@
+import memoize from 'lodash/memoize';
+
+import createEgoUtils from '@icgc-argo/ego-token-utils';
+import { getAppConfig } from '../../config';
+
+const TokenUtils = createEgoUtils(getAppConfig().EGO_PUBLIC_KEY);
+
+export const decodeToken = memoize((egoJwt?: string) =>
```
[the remark this answers is not in the document]
Comment (demariadaniel) [reply src/global/utils/egoJwt/index.ts:27]: Ask platform-ui: 
https://github.com/icgc-argo/platform-ui/blob/develop/global/utils/egoJwt/index.tsx#L32
https://github.com/icgc-argo/platform-ui/blob/develop/global/hooks/useAuthContext.tsx

Lots of utils rely on egoJwt being falsy, so RDPC likely will at one point as well

File: src/app/files/page.tsx [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+ */
+'use client';
+
+import Cookies from 'js-cookie';
+import Image from 'next/image';
+import { useEffect } from 'react';
+import urljoin from 'url-join';
+import { EGO_JWT_KEY } from '../../global/constants';
+import { getAppConfig } from '../../global/config';
+import { logOut } from '../../global/hooks/auth';
```
[the remark this answers is not in the document]
Comment (demariadaniel) [reply src/app/files/page.tsx]: Handy, updated

File: src/global/hooks/auth.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (ciaranschutte) [reply src/global/hooks/auth.tsx]: AuthContextValue is great. only preference is not prefixing types with `T_` or similar.

[the remark this answers is not in the document]
Comment (demariadaniel) [reply src/global/hooks/auth.tsx]: Updated

File: src/app/page.tsx [hunk 7]
```diff
@@ -69,6 +78,40 @@ const OvertureBanner: ComponentType = () => {
...
+							margin-right: 5px;
+						`}
+					/>
+					Login
+				</span>
+			</Button>
+		</a>
+	);
+};
+
```
[the remark this answers is not in the document]
Comment (demariadaniel) [reply src/app/page.tsx]: Yes I had tried that originally, updating now that TS issues are resolved

File: .env.schema (hunk 4 above)
[the remark this answers is not in the document]
Comment (demariadaniel) [reply .env.schema:10]: OK I'm updating `env.schema` to `env.local` and have removed publicRunTimeConfig

File: .gitignore [hunk 8]
```diff
@@ -25,12 +25,15 @@ yarn-debug.log*
 yarn-error.log*
 
 # local env files
-.env*.local
+.env
```
[the remark this answers is not in the document]
Comment (demariadaniel) [reply .gitignore:28]: To confirm: .env.local should be committed, and .gitignore should _not_ include .env*.local
Is that right?

File: next.config.js [hunk 9]
```diff
@@ -24,6 +24,11 @@ const nextConfig = {
...
 		emotion: true,
 	},
+	publicRuntimeConfig: {
+		EGO_API_ROOT: process.env.EGO_API_ROOT,
+		EGO_CLIENT_ID: process.env.EGO_CLIENT_ID,
+		EGO_PUBLIC_KEY: process.env.EGO_PUBLIC_KEY,
+	},
 };
 
 module.exports = nextConfig;
```
[the remark this answers is not in the document]
Comment (demariadaniel) [reply next.config.js:34]: This is removed, tested + working

File: .gitignore (hunk 8 above)
[the remark this answers is not in the document]
Comment (ciaranschutte) [reply .gitignore]: opposite.

`.env.local` should _never_ be commited, this was in the .gitignore already so you can probably just revert the commit to it. 

`.env.local` is what we're used to `.env` historically being
https://nextjs.org/docs/app/building-your-application/configuring/environment-variables#default-environment-variables

File: src/app/layout.tsx [hunk 10]
```diff
@@ -22,18 +22,25 @@
...
+	const loggedIn = Boolean(egoJwt && isValidJwt(egoJwt));
+
 	return (
 		<html lang="en">
 			<body className={workSans.className}>
 				<ThemeProvider>
-					<Header />
+					<Header loggedIn={loggedIn} />
 					{children}
 				</ThemeProvider>
```
Comment (ciaranschutte) [inline src/app/layout.tsx:45]: re slack message.

Issue is two fold:
1) `layout.tsx` is not re-rendered. there is no state change. acting as it should for SPA routing. something like the AuthProvider you had previously is the way to go I think, you can make a state change and the rest should re-render. very similar to platform-ui. I'd put it directly after `ThemeProvider` - github won't allow me to suggest commit on deleted lines - anyway - ThemeProvider will likely never re-render because we're not giving it dynamic themes where as AuthProvider. that is to say , context providers will re-render all children.  

the layout special convention component is similar to next12  `_app.tsx`
These two pages are worth a read:
- https://nextjs.org/docs/app/building-your-application/routing/pages-and-layouts
- https://nextjs.org/docs/app/api-reference/file-conventions ( some of these are VERY cool like `loading.js` )

2) we're in reactstrict mode. this means in dev renders will be called twice. its a way for react to call out things that may not be safe in production. https://react.dev/reference/react/StrictMode 
in this specific case, what if you make a network request to ego, then react re-renders the request page before you can act on the response? you get this exact double response happening. So it's making sure you handle these cases. we can bring in a library for all our network requests eg. react-query , otherwise I believe an AbortController is what you want for making sure duplicate requests are not sent

These seem like new additions above the scope of feedback so it might be best to just address PR feedback, merge, and then we can work through a more tightly scoped PR.

Comment (demariadaniel) [reply src/app/layout.tsx:48]: Alright, react-query handles the double request issue perfectly, and I've got the AuthProvider setup as you suggested and it's working better. The Login button show/hide is still not perfect but we can improve on it later.

**Comment by ciaranschutte** [comment]
@demariadaniel took a quick look. homepage login looks broken. I can share a screen on slack.

Works for a second. Then looks like a re-render happens and the <LoginButton> disappears. Bunch of hydration errors in the console.

File: src/app/layout.tsx [hunk 11]
```diff
@@ -38,7 +38,7 @@ export default function RootLayout({ children }: { children: ReactNode }) {
 	const path = usePathname();
 	const storedToken = getToken();
 	const [egoJwt, setEgoJwt] = useState(storedToken || '');
-	const initLoginState = path === '/logged-in' && !storedToken ? true : false;
+	const initLoginState = path === '/logged-in' && !(storedToken || egoJwt) ? true : false;
```
Comment (justincorrigible) [inline src/app/layout.tsx:41]: From the conditional logic here, I find quite hard to figure out what `initLoginState` as a boolean is supposed to mean. e.g. is it short for `wasUserLoggedInWhenComponentLoaded`? if so, why does having a token/jwt lead to `false`? the conditions read to me like a `wasUserNotLoggedInWhenComponentLoaded`, but I'm probably missing something here...

File: src/app/components/Header.tsx [hunk 12]
```diff
@@ -16,27 +16,44 @@
...
  */
+'use client';
 
-import { AppBar, css } from '@icgc-argo/uikit';
+import { AppBar, css, UserBadge } from '@icgc-argo/uikit';
 import Link from 'next/link';
 import Image from 'next/image';
-import argoLogo from '../../../public/assets/argo-logo.svg';
+import { getToken, useAuthContext } from '@/global/utils/auth';
+import argoLogo from '@/../public/assets/argo-logo.svg';
```
[the remark this answers is not in the document]
Comment (demariadaniel) [reply src/app/components/Header.tsx:25]: Haha yes I knew this was a bit wonky

File: src/global/config.ts [hunk 13]
```diff
@@ -16,11 +16,22 @@
  * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
  * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
+//'use client';
```
Comment (justincorrigible) [inline src/global/config.ts:19]: this will ideally not ever be necessary in this file
```suggestion
```

File: src/app/layout.tsx (hunk 11 above)
Comment (demariadaniel) [reply src/app/layout.tsx:41]: Fair, I'll update. I'm using this to manage the NavBar Login button / Profile badge state: 
- if you're on the homepage with no token, you should see Login
- if you're on the Logging In page you should see nothing
- if you have a token, you should see the Profile badge

Comment (justincorrigible) [reply src/app/layout.tsx:41]: All that from that little bit of logic? wow. 
💪 heavy lifting!

In that case, it almost seems like all this logic should be abstracted into the auth provider, and passed from there to the relevant children (as individual variables, like `shouldShowBadge`, `shouldRedirectToLogin`, etc.). 
Unless you foresee this being relevant at the layout base in the future?

File: src/app/components/Header.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (demariadaniel) [reply src/app/components/Header.tsx]: Updated to just use /public/assets

**Comment by ciaranschutte** [comment]
@demariadaniel these latest changes you're making look redundant when comparing to `develop` branch - be worth checking to make sure you have latest `develop` merged into this branch

**Comment by ciaranschutte** [comment]
@demariadaniel let me know when this is ready for review please - additional changes not related to feedback are hard to follow

File: src/global/utils/auth.tsx [hunk 14]
```diff
@@ -16,34 +16,71 @@
...
 
-export default function AuthProvider({ children }: { children: ReactNode }) {
-	const storedToken = Cookies.get(EGO_JWT_KEY);
+export function AuthProvider({ children }: { children: ReactNode }) {
+	const storedToken = getStoredToken();
 	const [egoJwt, setEgoJwt] = useState(storedToken || '');
-	const authData = { egoJwt, setEgoJwt };
-	return <AuthContext.Provider value={authData}>{children}</AuthContext.Provider>;
+	const path = usePathname();
+	const initLoginState = path === '/logging-in' && !egoJwt.length ? true : false;
```
Comment (ciaranschutte) [inline src/global/utils/auth.tsx:65]: I think @justincorrigible comment is good about changing the name because init does seem like initializing

```javascript
!egoJwt.length ?  true : false
```
this can just be the conditional `!egoJwt.legnth` - whenever a ternary is just returning `conditional ?  true : false` you can just do the conditional if that makes sense? Otherwise it's just if `true return true` if `false return false`

**Review by ciaranschutte — APPROVED** [review]
a few comments to keep an eye on for future PRs for this - approving to unblock

- the Cookie.set(token) and the egoJwt in state are somewhat coupled - there may be a need for a cleaner sync between cookie + sync - i think there may even be some code in platform-ui

- does setting the logging in state in AuthProvider re-render the logging - in page?

re our slack convo we can tackle these as we go - just mentioning them now

**Comment by demariadaniel** [comment]
> a few comments to keep an eye on for future PRs for this - approving to unblock
> 
> * the Cookie.set(token) and the egoJwt in state are somewhat coupled - there may be a need for a cleaner sync between cookie + sync - i think there may even be some code in platform-ui
> * does setting the logging in state in AuthProvider re-render the logging - in page?
> 
> re our slack convo we can tackle these as we go - just mentioning them now

Thanks @ciaranschutte I have fixed the last few changes as part of https://github.com/icgc-argo/rdpc-ui/pull/39
